crates/typst/src/layout/inline/mod.rs

```diff
@@ -1085,7 +1085,9 @@ fn line<'a>(
                         // Compression of punctuation marks at line start or line end
                         let shrink_amount = last_glyph.shrinkability().1;
                         let punct = reshaped.glyphs.to_mut().last_mut().unwrap();
-                        punct.shrink_right(shrink_amount);
+                        // No need to add `stretchability` of the last CJK punctuation glyph to avoid stretching in line adjustment.
+                        punct.x_advance -= shrink_amount;
+                        punct.adjustability.shrinkability.1 -= shrink_amount;
```

Comment:
There are three places of adjustments:

- At line start;
- At line end;
- Between two consecutive punctuation marks.

All of the three places shouldn't add the stretchability. So maybe you can just change the implementation of `shrink_right` and `shrink_left`.

Reply:
Ok, I'll check it, and update this PR, Thanks.

Reply:
@peng1999 I have modified `shrink_right` and `shrink_left`, and it works.